xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/skin/SkinEnvironmentResource.java

```diff
@@ -59,6 +72,7 @@ public String getURL(boolean forceSkinAction) throws Exception
             url = urlf.createSkinURL(this.resourceName, getRepository().getId(), xcontext);
         }
 
+        url = urlf.addQueryParameters(url, parameters);
```

Comment:
I still think this have nothing to do with XWikiURLFactory. Sounds more like something for a generic URL related tool on xwiki-commons side (did not saw anything in Apache commons but maybe there is).

Reply:
Actually I'm already using an Apache commons (https://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/client/utils/URIBuilder.html) but I still need to encapsulate it in a method since I need to call the builder, encode the parameters etc. 

I checked in xwiki-commons I don't really where to put that in the existing modules. Were you thinking about creating a new module specifically for that? Seems a bit overkill.

Reply:
My main point is that it does not make any sense as a XWikiURLFactory method and should go in some URL related tool instead.

Ideally it should probably go in some xwiki-commons-lang and rename xwiki-commons-text). In the meantime I guess the easiest and safest in term of technical debt is to just add some URLUtils in an internal package of oldcore or xwiki-platform-url-api for example.

Reply:
So, I agree with you that the utility method shouldn't be added in XWikiURLFactory. Now I really don't want to create a specific class URLUtils just to add a new method we only use twice. Moreover the purpose of XWikiURLFactory is to built dedicated URL for different stuff. 

Finally I propose to add some new methods in XWikiURLFactory by adding the queryParameters parameter to some existing method such as `createSkinURL` and `createResourceURL`, so when it creates those URLs it will be able to also put the needed query parameters. It avoid to create an URL and to modify it afterwards. WDYT?

Reply:
> Finally I propose to add some new methods in XWikiURLFactory by adding the queryParameters parameter to some existing method such as `createSkinURL` and `createResourceURL`, so when it creates those URLs it will be able to also put the needed query parameters. It avoid to create an URL and to modify it afterwards. WDYT?

I'm fine with that, I remember suggesting it in a previous comment.